common/quotas/global/collection/internal/atomicmap_external_test.go

```diff
@@ -0,0 +1,185 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package internal_test
+
+import (
+	"math/rand"
+	"runtime"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/uber/cadence/common/quotas/global/collection/internal"
+)
+
+func TestMapBasics(t *testing.T) {
+	const (
+		loaded1 = "four"
+		loaded2 = "tenletters"
+		tried   = "should not exist"
+	)
+	type custom struct{ value int }
+
+	assertContentsEqual := func(t *testing.T, m *internal.AtomicMap[string, *custom], expected map[string]int) {
+		dup := make(map[string]int) // to avoid mutating the original
+		for k, v := range expected {
+			dup[k] = v
+		}
+		m.Range(func(k string, v *custom) bool {
+			if _, ok := dup[k]; ok {
+				delete(dup, k)
+			} else {
+				t.Errorf("ranged over unexpected or duplicate key %q", k)
+			}
+			return true
+		})
+		assert.Empty(t, dup, "did not find some contents")
+	}
+
+	m := internal.NewAtomicMap(func(key string) *custom {
+		return &custom{value: len(key)}
+	})
+
+	t.Run("load should work", func(t *testing.T) {
+		v := m.Load(loaded1)
+		assert.Equal(t, len(loaded1), v.value, "should use the constructed value initially, not a zero value")
+		v.value = 10
+		reload := m.Load(loaded1)
+		assert.Equal(t, v, reload, "should return the same object when loaded more than once")
+	})
+	t.Run("range should walk over only the one key", func(t *testing.T) {
+		assertContentsEqual(t, m, map[string]int{
+			loaded1: len(loaded1),
+		})
+	})
+	t.Run("loading a second value should range over two", func(t *testing.T) {
+		// init this value too
+		v2 := m.Load(loaded2)
+		assert.Equal(t, len(loaded2), v2.value, "sanity check: loaded2 should be created correctly, like loaded1")
+		assertContentsEqual(t, m, map[string]int{
+			loaded1: len(loaded1),
+			loaded2: len(loaded2),
+		})
+	})
+}
+
+func TestMapNotRacy(t *testing.T) {
+	creates := atomic.NewInt64(0)
+	// using a string pointer just to make things a bit riskier / more sensitive to races since mutation is possible.
+	// no mutation currently occurs, but it seems slightly safer to leave it here for future changes.
+	m := internal.NewAtomicMap(func(key string) *string {
+		s := key
+		s += "-"
+		s += strconv.Itoa(int(creates.Inc())) // just to be recognizable
+		return &s
+	})
+
+	// call ALL the methods concurrently
+	var g errgroup.Group
+	const loops = 1000 // 100 has had some coverage flapping, raise further if needed
```

Comment:
nit: mismatch

Reply:
I can reword that, it was meant as a "using 1000 because 100 was too low"